README.md
Grey codes a.k.a. Reflected binary codes in C, made simple
===============================================================================

Grey codes or Reflected binary codes are encodings of unsigned integers
such that two adjacent values have only 1 bit of difference. Originally
designed for electromechanical switches so that the state would transition
cleanly from one value to the other without other values being represented
in between. More details on 
[Wikipedia](https://en.wikipedia.org/wiki/Gray_code).

This is a tiny tiny C99 library that converts unsigned integers of any size
from and to Grey codes.

By default, the library will operate on `uint64_t` integers for conversions
from/to Grey codes. If you prefer using smaller integer (but also limit the 
domain of the Grey codes and their values), redefine the macro `GREY_UINTBITS`
to 32, 16 or 8 instead of 64.


Usage example
----------------------------------------

```c
// Convert to Grey code
grey_code_t my_code = grey_to(10); // my_code is now 0b1111 = 0x0F = 15

// Convert from Grey code
grey_int_t my_value = grey_from(my_code);  // my_value is now 10 again

// Handy macro for incrementing Grey codes
grey_code_t my_code = grey_to(103);  // my_code is now 84, representing 103
my_code = grey_incr(my_value);  // my_code is now 92, representing 104
my_value = grey_from(my_code);  // my_value is now 103

// Or for decrementing
my_code = grey_decr(85);

// Or for custom addition/subtraction (may be unsafe!)
my_code = grey_add(85, 5);

// Printing format
my_code = grey_to(42);
printf("Int %u "
       "= Grey decimal %" GREY_FMT
       " = Grey lowercase hex 0x%" GREY_FMTx
       " = Grey uppercase hex 0x%" GREY_FMTX "\n",
       42, my_code, my_code, my_code);

// Converting to binary string
char binstr[GREY_UINTBITS + 1];
uint8_t len = grey_binstr(binstr, 7);  // str now contains "111\0", len is 3
```

You can also check the `tst/test.c` file for more examples.




Include it in your project
----------------------------------------

### Static source inclusion

Copy the `inc/grey.h` and `src/grey.c` files into your existing
C project, add them to the source folders and compile. Done.

If you prefer using smaller integers, redefine `GREY_UINTBITS`.


### Compiling into all possible targets

```
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DGREY_UINTBITS=64
cmake --build .
```

This will build all targets:

- a `libgrey.dylib`/`libgrey.dll` shared library
- a `libgreystatic.a` static library
- a test runner executable `test_grey`
- the Doxygen documentation (if Doxygen is installed)

To compile with the optimisation for size, use the
`-DCMAKE_BUILD_TYPE=MinSizeRel` flag instead.

If you prefer using smaller integers, set `-DGREY_UINTBITS=32` (or 16 or 8).


CHANGELOG.md
Changelog
===============================================================================

All notable changes to this project will be documented in this file.

The format is based on
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

*******************************************************************************

[1.0.0] - 2020-04-11
----------------------------------------

Initial version.


### Added

- Conversion to and from Grey codes
- Parametric underlying integer type
- Simple conversion to binary string
- Static, shared library build
- Test runner build
- Doxygen build


inc/grey.h
/**
 * @file
 *
 * Grey codes a.k.a. Reflected binary codes, made simple.
 *
 * This is a tiny tiny C99 library that converts unsigned integers of any size
 * from and to Grey codes.
 *
 * By default, the library will operate on `uint64_t` integers for conversions
 * from/to Grey codes. If you prefer using smaller integer (but also limit the
 * domain of the Grey codes and their values), redefine the macro
 * `GREY_UINTBITS` to 32, 16 or 8 instead of 64.
 *
 * @copyright Copyright © 2020, Matjaž Guštin <dev@matjaz.it>
 * <https://matjaz.it>. All rights reserved.
 * @license BSD 3-clause license.
 */

#ifndef GREY_H
#define GREY_H

#ifdef __cplusplus
extern "C"
{
#endif

/**
 * Version of the grey.h API using semantic versioning.
 */
#define GREY_VERSION "1.0.0"

#include <stdint.h>
#include <inttypes.h>

/**
 * @property GREY_UINTBITS
 * Defines the domain in bits of the Grey codes. In other words what is the
 * max value a Grey code could hold.
 *
 * - Set it to 64 to convert `uint64_t` values to and from Grey codes.
 * - Set it to 32 to convert `uint32_t` values to and from Grey codes.
 * - Set it to 16 to convert `uint16_t` values to and from Grey codes.
 * - Set it to  8 to convert `uint8_t`  values to and from Grey codes.
 *
 * #grey_code_t and #grey_int_t will automatically be defined as the proper
 * integer types to hold the Grey-encoded values and their binary (regular
 * integer) values.
 *
 * Defaults to 64 thus using `uint64_t` integers only, unless specified
 * otherwise at compile time.
 */
#ifndef GREY_UINTBITS
#define GREY_UINTBITS 64
#endif

/**
 * @property grey_code_t
 * Grey-encoded value, to distinguish them from binary values.
 *
 * Can be altered with #GREY_UINTBITS.
 */
/**
 * @property GREY_MAX
 * Maximum allowed value that #grey_code_t and #grey_int_t support.
 *
 * Can be increased with #GREY_UINTBITS.
 */
/**
 * @property GREY_FMT
 * Format specifier for `printf` and similar functions of the
 * decimal integer representation of #grey_code_t and #grey_int_t.
 *
 * Usage example:
 *
 *     grey_code_t code = 15U;
 *     printf("My Grey code is " GREY_FMT "!\n", code); // prints 15
 */
/**
 * @property GREY_FMTx
 * Format specifier for `printf` and similar functions of the
 * hexadecimal lowercase integer representation of #grey_code_t and
 * #grey_int_t.
 *
 * Usage example:
 *
 *     grey_code_t code = 15;
 *     printf("My Grey code is " GREY_FMTx "!\n", code); // prints "f"
 */
/**
 * @property GREY_FMTX
 * Format specifier for `printf` and similar functions of the
 * hexadecimal uppercase integer representation of #grey_code_t and
 * #grey_int_t.
 *
 * Usage example:
 *
 *     grey_code_t code = 15;
 *     printf("My Grey code is " GREY_FMTX "!\n", code); // prints "F"
 */
#if (GREY_UINTBITS == 64)
typedef uint64_t grey_code_t;
#define GREY_MAX ((uint64_t) UINT64_MAX)
#define GREY_FMT PRIu64
#define GREY_FMTx PRIx64
#define GREY_FMTX PRIX64
#elif (GREY_UINTBITS == 32)
typedef uint32_t grey_code_t;
#define GREY_MAX ((uint32_t) UINT32_MAX)
#define GREY_FMT PRIu32
#define GREY_FMTx PRIx32
#define GREY_FMTX PRIX32
#elif (GREY_UINTBITS == 16)
typedef uint16_t grey_code_t;
#define GREY_MAX ((uint16_t) UINT16_MAX)
#define GREY_FMT PRIu16
#define GREY_FMTx PRIx16
#define GREY_FMTX PRIX16
#elif (GREY_UINTBITS == 8)
typedef uint8_t grey_code_t;
#define GREY_MAX ((uint8_t) UINT8_MAX)
#define GREY_FMT PRIu8
#define GREY_FMTx PRIx8
#define GREY_FMTX PRIX8
#else
#error "Please set GREY_UINTBITS to a value in {8, 16, 32, 64}."
#endif

/** Binary value (regular integer), of the same size as Grey-coded values. */
typedef grey_code_t grey_int_t;

/**
 * Utility wrapper adding/subtracting a delta to a Grey-encoded value.
 *
 * @warning No checks are performed for overflows/underflows/integer conversion
 * errors etc.!
 * @param grey value to increase/decrease.
 * @param delta value to add/remove from the grey code, signed.
 * @return increased/decreased Grey code.
 */
#define grey_add(grey, delta) grey_to(grey_from((grey)) + (delta))

/**
 * Utility wrapper incrementing a Grey-encoded value by 1.
 *
 * @warning No checks are performed for overflows!
 * @param grey value to increment.
 * @return incremented Grey code = `grey+1`.
 */
#define grey_incr(grey) grey_add((grey), 1)

/**
 * Utility wrapper decrementing a Grey-encoded value by 1.
 *
 * @warning No checks are performed for underflows!
 * @param grey value to decrement.
 * @return decremented Grey code = `grey-1`.
 */
#define grey_decr(grey) grey_add((grey), -1)

/**
 * Converts a regular binary unsigned integer to Grey code.
 *
 * @param value the binary value (regular integer) to convert
 * @return \p value converted into Grey code
 */
grey_code_t grey_to(grey_int_t value);

/**
 * Converts a Grey-encoded value into a regular binary unsigned integer.
 *
 * @param grey value to convert
 * @return binary value (regular integer) of the Grey code
 */
grey_int_t grey_from(grey_code_t grey);


/**
 * Fills a string with the Grey code in binary representation.
 *
 * The representation is:
 * - ASCII
 * - big endian (as humans would read and write the number)
 * - contains no spaces
 * - is **null-terminated**
 * - is at most #GREY_UINTBITS bytes in size (**excluding** the
 *   null-terminator).
 *
 * Example:
 *     `grey_binstr(str, 0x0E)` fills `str` with `"1110\0"` and returns `4`.
 *
 * Note: this is just a utility function, is not meant to be super-efficient.
 *
 * @param[out] str buffer of #GREY_UINTBITS+1 bytes (the "+1" is the space
 *             for the null-terminator).
 * @param[in] grey value to encode.
 * @returns length of the binary string written into \p str **excluding** the
 *          null-terminator
 */
uint8_t grey_binstr(char str[GREY_UINTBITS+1], grey_code_t grey);

#ifdef __cplusplus
}
#endif

#endif  /* GREY_H */
